### PR TITLE
SSCS-10644-Linked-Cases-Update : fix only one case ref showing

### PR DESCRIPTION
**Pull request by Jonathan-Weston1**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/SSCS-10644

### Change description ###
Update service Linked Hearings endpoint to return individual caseReferences

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/LinkedCasesMapping.java [hunk 1]
```diff
@@ -20,17 +23,66 @@ private LinkedCasesMapping() {
...
-                .filter(Objects::nonNull)
-                .map(CaseLinkDetails::getCaseReference)
-                .filter(StringUtils::isNotBlank)
-                .map(caseReference -> ServiceLinkedCases.builder()
-                    .caseReference(caseReference)
-                    .caseName(HearingsCaseMapping.getPublicCaseName(caseData))
-                    .reasonsForLink(HearingsCaseMapping.getReasonsForLink(caseData))
-                    .build())
-                .collect(Collectors.toList());
+            .orElseGet(Collections::emptyList).stream()
```
Comment (dattipoeHMCTS) [inline src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/LinkedCasesMapping.java:26]: I think having this here will work but you should be able to move the or else to the bottom. I believe ofNullable will falldown to the orElse once it hits a null value. Same as below

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/LinkedCasesMapping.java [hunk 2]
```diff
@@ -20,17 +23,66 @@ private LinkedCasesMapping() {
...
+
+        if (linkedReferences.isEmpty()) {
+            return Collections.emptyList();
+        }
+
+        List<ServiceLinkedCases> serviceLinkedCases = new ArrayList<>();
+        ServiceLinkedCases linkedCase;
+        for (String linkRef : linkedReferences) {
+            if (linkRef != null) {
+                //suppressed warning
```
Comment (dattipoeHMCTS) [inline src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/LinkedCasesMapping.java:58]: Comment

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/LinkedCasesMapping.java [hunk 3]
```diff
@@ -20,17 +23,66 @@ private LinkedCasesMapping() {
...
+            .collect(Collectors.toList());
+
+        if (linkedReferences.isEmpty()) {
+            return Collections.emptyList();
+        }
+
+        List<ServiceLinkedCases> serviceLinkedCases = new ArrayList<>();
+        ServiceLinkedCases linkedCase;
+        for (String linkRef : linkedReferences) {
+            if (linkRef != null) {
```
Comment (dattipoeHMCTS) [inline src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/LinkedCasesMapping.java:57]: Is this possible with the above stream?

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/LinkedCasesMapping.java [hunk 4]
```diff
@@ -20,17 +23,66 @@ private LinkedCasesMapping() {
...
+            .filter(Objects::nonNull)
+            .map(CaseLinkDetails::getCaseReference)
+            .collect(Collectors.toList());
+
+        if (linkedReferences.isEmpty()) {
+            return Collections.emptyList();
+        }
+
+        List<ServiceLinkedCases> serviceLinkedCases = new ArrayList<>();
+        ServiceLinkedCases linkedCase;
```
Comment (dattipoeHMCTS) [inline src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/LinkedCasesMapping.java:55]: You should be able to move this into the for loop

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/LinkedCasesMapping.java [hunk 5]
```diff
@@ -20,17 +23,66 @@ private LinkedCasesMapping() {
...
+        return serviceLinkedCases;
+    }
+
+    private static SscsCaseData getLinkedCaseData(SscsCaseData caseData, CcdCaseService ccdCaseService, String caseReference) throws GetCaseException {
+        List<CaseLink> caseLink = caseData.getLinkedCase();
+
+        String linkedReference = null;
+
+        for (CaseLink link : caseLink) {
+            if (link != null && link.getValue() != null && link.getValue().getCaseReference() != null && link.getValue().getCaseReference().equals(caseReference)) {
```
Comment (dattipoeHMCTS) [inline src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/LinkedCasesMapping.java:79]: I'd hope none of these could be null but it would be nicer to use an Optional.ofNullable here too

File: src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/LinkedCasesMappingTest.java [hunk 6]
```diff
@@ -1,58 +1,91 @@
...
-    public static final long CASE_ID = 99250807409918L;
+    @Mock
+    private CcdService ccdService;
+
+    @Mock
+    IdamService idamService;
+
+    @BeforeEach
+    void setUp() {
+        openMocks(this);
```
Comment (dattipoeHMCTS) [inline src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/LinkedCasesMappingTest.java:37]: I'd use

`
@InjectMocks
CcdCaseService ccdCaseService
`
At class level instead of openmocks

**Review by dattipoeHMCTS — CHANGES_REQUESTED** [review]
Is there a test case to make sure the issue which caused the reopen is covered? You might also need some test cases to cover null paths

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/LinkedCasesMapping.java (hunk 1 above)
Comment (Jonathan-Weston1) [reply src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/LinkedCasesMapping.java:26]: Discussed on a call. Happy to close

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/LinkedCasesMapping.java (hunk 3 above)
Comment (Jonathan-Weston1) [reply src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/LinkedCasesMapping.java:56]: Discussed on a call happy to close

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/LinkedCasesMapping.java [hunk 7]
```diff
@@ -20,17 +23,68 @@ private LinkedCasesMapping() {
...
+        List<ServiceLinkedCases> serviceLinkedCases = new ArrayList<>();
+        for (String linkRef : linkedReferences) {
+            if (linkRef != null) {
+                ServiceLinkedCases linkedCase = new ServiceLinkedCases();
+                linkedCase.setCaseReference(linkRef);
+                linkedCase.setCaseName(HearingsCaseMapping.getPublicCaseName(getLinkedCaseData(
+                    caseData,
+                    ccdCaseService,
+                    linkRef
+                )));
```
[the remark this answers is not in the document]
Comment (TeaWithLucas) [reply src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/LinkedCasesMapping.java:63]: can you put a null check here instead of changing getPublicCaseName?

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/LinkedCasesMapping.java [hunk 8]
```diff
@@ -20,17 +23,68 @@ private LinkedCasesMapping() {
...
+            .filter(StringUtils::isNotBlank)
+            .map(caseReference -> ServiceLinkedCases.builder()
+                .caseReference(caseReference)
+                .caseName(HearingsCaseMapping.getPublicCaseName(caseData))
+                .reasonsForLink(HearingsCaseMapping.getReasonsForLink(caseData))
+                .build())
+            .collect(Collectors.toList());
+    }
+
+    @SuppressWarnings("PMD.AvoidInstantiatingObjectsInLoops")
```
Comment (TeaWithLucas) [inline src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/LinkedCasesMapping.java:40]: Can you address this instead of suppressing?

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/LinkedCasesMapping.java [hunk 9]
```diff
@@ -20,17 +23,68 @@ private LinkedCasesMapping() {
...
+    public static List<ServiceLinkedCases> getLinkedCasesWithNameAndReasons(SscsCaseData caseData, CcdCaseService ccdCaseService) throws GetCaseException {
```
Comment (TeaWithLucas) [inline src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/LinkedCasesMapping.java:41]: This method feels like it should be split into two, one for getting linkedReferences, one for creating serviceLinkedCases

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/LinkedCasesMapping.java [hunk 10]
```diff
@@ -20,17 +23,68 @@ private LinkedCasesMapping() {
...
+            .collect(Collectors.toList());
+
+        if (linkedReferences.isEmpty()) {
+            return Collections.emptyList();
+        }
+
+        List<ServiceLinkedCases> serviceLinkedCases = new ArrayList<>();
+        for (String linkRef : linkedReferences) {
+            if (linkRef != null) {
+                ServiceLinkedCases linkedCase = new ServiceLinkedCases();
```
Comment (TeaWithLucas) [inline src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/LinkedCasesMapping.java:57]: Can you use the builder instead of creating a new instance

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsCaseMapping.java [hunk 11]
```diff
@@ -25,6 +25,7 @@
 public final class HearingsCaseMapping {
 
     public static final String CASE_DETAILS_URL = "%s/cases/case-details/%s";
+    public static final String EMPTY_STRING = "";
```
Comment (TeaWithLucas) [inline src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsCaseMapping.java:28]: (minor) if not being used, can you remove?

**Review by dattipoeHMCTS — APPROVED** [review]
